cggmp21/src/signing.rs

```diff
@@ -987,93 +1114,86 @@ where
-            &msg.psi.1,
-        )
-        .err();
-
-        tracer.stage("Validate hat_psi");
-        let hat_psi_invalid = pi_aff::non_interactive::verify::<E, D>(
-            &unambiguous::ProofPsi {
-                sid,
-                prover: j,
-                hat: true,
-            },
-            &R_i.into(),
-            pi_aff::Data {
-                key0: &dec_i,
-                key1: &enc_j,
-                c: &K_i,
-                d: &msg.hat_D,
-                y: &msg.hat_F,
-                x: &X_j,
-            },
-            &msg.hat_psi.0,
-            &security_params.pi_aff,
-            &msg.hat_psi.1,
-        )
-        .err();
-
-        tracer.stage("Validate psi_prime");
-        let psi_prime_invalid = pi_log::non_interactive::verify::<E, D>(
-            &unambiguous::ProofLog {
-                sid,
-                prover: j,
-                prime_prime: false,
-            },
-            &R_i.into(),
-            pi_log::Data {
-                key0: &enc_j,
-                c: &ciphertexts.G,
-                x: &msg.Gamma,
-                b: &Point::<E>::generator().to_point(),
-            },
-            &msg.psi_prime.0,
-            &security_params.pi_log,
-            &msg.psi_prime.1,
-        )
-        .err();
-
-        if psi_invalid.is_some() || hat_psi_invalid.is_some() || psi_prime_invalid.is_some() {
-            faulty_parties.push((
-                j,
-                ciphertext_msg_id,
-                msg_id,
-                (psi_invalid, hat_psi_invalid, psi_prime_invalid),
-            ))
-        }
-        runtime.yield_now().await;
-    }
+    let faulty_parties =
+        utils::collect_blame_with_data(&round2_msgs, &ciphertexts, |j, msg, ciphertexts| {
+            tracer.stage("Retrieve auxiliary data");
+            let X_j = X[usize::from(j)];
+            let enc_j = &N[usize::from(j)];
+
+            tracer.stage("Validate tilde_psi_j");
+            let tilde_psi_invalid = pi_elog::non_interactive::verify::<E, D>(
+                &unambiguous::ProofElog {
+                    sid,
+                    prover: j,
+                    prime: false,
+                },
+                pi_elog::Data {
+                    l: &ciphertexts.B1,
+                    m: &ciphertexts.B2,
+                    x: &ciphertexts.Y,
+                    y: &msg.Gamma,
+                    h: &Point::generator().to_point(),
+                },
+                &msg.tilde_psi.0,
+                &msg.tilde_psi.1,
+            )
+            .err();
+
+            tracer.stage("Validate psi_i,j");
+            let psi_j_invalid = pi_aff::non_interactive::verify::<E, D>(
+                &unambiguous::ProofPsi {
+                    sid,
+                    prover: j,
+                    hat: false,
+                },
+                &R_i.into(),
+                pi_aff::Data {
+                    key_j: &dec_i,
+                    key_i: enc_j,
```

Comment:
I'm unsure about these parameters because above they were inverted.

Just underline this for you to double check if there is an inversion or if it's totally fine.

Reply:
Actually, it's structure fields, the order in which they are specified doesn't matter, you can swap those lines, nothing will change

Reply:
All good then 😉